src/lib.rs

```diff
@@ -715,251 +738,307 @@ enum Decoder {
             },
-            Expr::Record(fields) => {
-                Value::record(fields.iter().map(|(label, expr)| (label, expr.eval(scope))))
+            Expr::Record(fields) => Cow::Owned(Value::record(
+                fields
+                    .iter()
+                    .map(|(label, expr)| (label, expr.eval(scope).into_owned())),
+            )),
+            Expr::RecordProj(head, label) => match head.eval(scope) {
+                Cow::Owned(v) => Cow::Owned(v.record_proj(label).clone()),
+                Cow::Borrowed(v) => Cow::Borrowed(v.record_proj(label)),
+            },
+            Expr::Variant(label, expr) => {
+                Cow::Owned(Value::variant(label, expr.eval(scope).into_owned()))
             }
-            Expr::RecordProj(head, label) => head.eval(scope).record_proj(label),
-            Expr::Variant(label, expr) => Value::variant(label, expr.eval(scope)),
-            Expr::Seq(exprs) => Value::Seq(exprs.iter().map(|expr| expr.eval(scope)).collect()),
+            Expr::Seq(exprs) => Cow::Owned(Value::Seq(
+                exprs
+                    .iter()
+                    .map(|expr| expr.eval(scope).into_owned())
+                    .collect(),
+            )),
             Expr::Match(head, branches) => {
                 let head = head.eval(scope);
-                let initial_len = scope.len();
-                let (_, expr) = branches
-                    .iter()
-                    .find(|(pattern, _)| head.matches(scope, pattern))
-                    .expect("exhaustive patterns");
-                let value = expr.eval(scope);
-                scope.truncate(initial_len);
-                value
+                for (pattern, expr) in branches {
+                    let mut pattern_scope = Scope::new();
+                    if head.matches(&mut pattern_scope, pattern) {
+                        let initial_len = scope.len();
+                        scope.extend(pattern_scope);
+                        let value = expr.eval(scope).into_owned();
+                        scope.truncate(initial_len);
+                        return Cow::Owned(value);
+                    }
+                }
+                panic!("non-exhaustive patterns");
             }
             Expr::Lambda(_, _) => panic!("cannot eval lambda"),
 
             Expr::BitAnd(x, y) => match (x.eval_value(scope), y.eval_value(scope)) {
-                (Value::U8(x), Value::U8(y)) => Value::U8(x & y),
-                (Value::U16(x), Value::U16(y)) => Value::U16(x & y),
-                (Value::U32(x), Value::U32(y)) => Value::U32(x & y),
+                (Value::U8(x), Value::U8(y)) => Cow::Owned(Value::U8(x & y)),
+                (Value::U16(x), Value::U16(y)) => Cow::Owned(Value::U16(x & y)),
+                (Value::U32(x), Value::U32(y)) => Cow::Owned(Value::U32(x & y)),
                 (x, y) => panic!("mismatched operands {x:?}, {y:?}"),
             },
```

Comment:
```suggestion
            Expr::BitAnd(x, y) => (match (x.eval_value(scope), y.eval_value(scope)) {
                (Value::U8(x), Value::U8(y)) => Value::U8(x & y),
                (Value::U16(x), Value::U16(y)) => Value::U16(x & y),
                (Value::U32(x), Value::U32(y)) => Value::U32(x & y),
                (x, y) => panic!("mismatched operands {x:?}, {y:?}"),
            }).into(),
```

or

```suggestion
            Expr::BitAnd(x, y) => Cow::Owned(match (x.eval_value(scope), y.eval_value(scope)) {
                (Value::U8(x), Value::U8(y)) => Value::U8(x & y),
                (Value::U16(x), Value::U16(y)) => Value::U16(x & y),
                (Value::U32(x), Value::U32(y)) => Value::U32(x & y),
                (x, y) => panic!("mismatched operands {x:?}, {y:?}"),
            }),
```

might be a bit more compact, possibly less legible though...

Reply:
That's good I think!

Reply:
Done.